File: app_manager/appmanager/src/main/java/app/editors/manager/mvp/presenters/main/DocsWebDavPresenter.java
package app.editors.manager.mvp.presenters.main;

import android.annotation.SuppressLint;
import android.content.ClipData;
import android.net.Uri;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Date;
import java.util.List;

import app.editors.manager.R;
import app.editors.manager.app.Api;
import app.editors.manager.app.App;
import app.editors.manager.app.WebDavApi;
import app.editors.manager.managers.providers.BaseFileProvider;
import app.editors.manager.managers.providers.WebDavFileProvider;
import app.editors.manager.mvp.models.account.AccountsSqlData;
import app.editors.manager.mvp.models.explorer.Explorer;
import app.editors.manager.mvp.models.explorer.File;
import app.editors.manager.mvp.models.explorer.Folder;
import app.editors.manager.mvp.models.explorer.Item;
import app.editors.manager.mvp.models.models.ModelExplorerStack;
import app.editors.manager.mvp.models.request.RequestCreate;
import app.editors.manager.mvp.views.main.DocsWebDavView;
import app.editors.manager.ui.dialogs.ContextBottomDialog;
import app.editors.manager.ui.views.custom.PlaceholderViews;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import lib.toolkit.base.managers.receivers.ExportReceiver;
import lib.toolkit.base.managers.utils.ContentResolverUtils;
import lib.toolkit.base.managers.utils.FileUtils;
import lib.toolkit.base.managers.utils.NetworkUtils;
import lib.toolkit.base.managers.utils.PermissionUtils;
import lib.toolkit.base.managers.utils.StringUtils;
import lib.toolkit.base.managers.utils.TimeUtils;
import moxy.InjectViewState;

@InjectViewState
public class DocsWebDavPresenter extends DocsBasePresenter<DocsWebDavView> {

    public static final String TAG = DocsWebDavPresenter.class.getSimpleName();

    private File mTempFile;

    private ExportReceiver mExportReceiver;
    private Disposable mDownloadDisposable;

    public DocsWebDavPresenter() {
        App.getApp().getAppComponent().inject(this);
        mToken = mPreferenceTool.getToken();
        mModelExplorerStack = new ModelExplorerStack();
        mFilteringValue = "";
        mPlaceholderType = PlaceholderViews.Type.NONE;
        mIsContextClick = false;
        mIsFilteringMode = false;
        mIsSelectionMode = false;
        mIsFoldersMode = false;
        mExportReceiver = new ExportReceiver();
        mFileProvider = getProvider();
    }

    private BaseFileProvider getProvider() {
        AccountsSqlData accountsSqlData = mAccountSqlTool.getAccountOnline();
        if (accountsSqlData != null) {
            return new WebDavFileProvider(WebDavApi.getApi(accountsSqlData.getScheme() + accountsSqlData.getPortal()),
                    WebDavApi.Providers.valueOf(accountsSqlData.getWebDavProvider()));
        } else {
            return null;
        }
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        mExportReceiver.setOnExportReceiver(uri -> upload(uri, null));
        mContext.registerReceiver(mExportReceiver, ExportReceiver.getFilters());
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mExportReceiver.setOnExportReceiver(null);
        mContext.unregisterReceiver(mExportReceiver);
    }

    @Override
    public void getNextList() {
        // Stub
    }

    @Override
    public void createDocs(@NonNull String title) {
        final String id = mModelExplorerStack.getCurrentId();
        if (id != null) {
            final RequestCreate requestCreate = new RequestCreate();
            requestCreate.setTitle(title);
            mDisposable.add(mFileProvider.createFile(id, requestCreate).subscribe(file -> {
                addFile(file);
                setPlaceholderType(PlaceholderViews.Type.NONE);
                getViewState().onDialogClose();
                getViewState().onOpenLocalFile(file);
            }, this::fetchError));

            showDialogWaiting(TAG_DIALOG_CANCEL_SINGLE_OPERATIONS);
        }
    }

    @Override
    public void getFileInfo() {
        if (mItemClicked != null && mItemClicked instanceof File) {
            final File file = (File) mItemClicked;
            final String extension = file.getFileExst();

            if (StringUtils.isImage(extension)) {
                addRecent(file);
                getViewState().onFileMedia(removeVideo(getListMedia(file.getId())), true);
                return;
            }
        }

        showDialogWaiting(TAG_DIALOG_CANCEL_UPLOAD);
        mDownloadDisposable = mFileProvider.fileInfo(mItemClicked)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(file -> {
                            mTempFile = file;
                            getViewState().onDialogClose();
                            getViewState().onOpenLocalFile(file);
                        }
                        , this::fetchError);
    }

    @Override
    public void addRecent(@NonNull File file) {
        AccountsSqlData account = mAccountSqlTool.getAccountOnline();
        if (account != null) {
            if (StringUtils.isImage(file.getFileExst())) {
                mAccountSqlTool.addRecent(file.getId(), file.getWebUrl(), file.getTitle(), file.getPureContentLength(),
                        false,true, new Date(), account);
            } else {
                mAccountSqlTool.addRecent(file.getViewUrl(), file.getWebUrl(), file.getTitle(), file.getPureContentLength(),
                        false,true, new Date(), account);
            }

        }
//        if (mTempFile != null && !mTempFile.getWebUrl().equals("")) {
//            FileUtils.asyncDeletePath(Uri.parse(mTempFile.getWebUrl()).getPath());
//        }
    }

    @Override
    protected void updateViewsState() {
        if (mIsSelectionMode) {
            getViewState().onStateUpdateSelection(true);
            getViewState().onActionBarTitle(String.valueOf(mModelExplorerStack.getCountSelectedItems()));
            getViewState().onStateAdapterRoot(mModelExplorerStack.isNavigationRoot());
            getViewState().onStateActionButton(false);
        } else if (mIsFilteringMode) {
            getViewState().onActionBarTitle(mContext.getString(R.string.toolbar_menu_search_result));
            getViewState().onStateUpdateFilter(true, mFilteringValue);
            getViewState().onStateAdapterRoot(mModelExplorerStack.isNavigationRoot());
            getViewState().onStateActionButton(false);
        } else if (!mModelExplorerStack.isRoot()) {
            getViewState().onStateAdapterRoot(false);
            getViewState().onStateUpdateRoot(false);
            getViewState().onStateActionButton(true);
            getViewState().onActionBarTitle(getCurrentTitle());
        } else {
            if (mIsFoldersMode) {
                getViewState().onActionBarTitle(mContext.getString(R.string.operation_title));
                getViewState().onStateActionButton(false);
            } else {
                getViewState().onActionBarTitle("");
                getViewState().onStateActionButton(true);
            }

            getViewState().onStateAdapterRoot(true);
            getViewState().onStateUpdateRoot(true);
        }
    }

    @Override
    public void onContextClick(Item item, int position, boolean isTrash) {
        onClickEvent(item, position);
        mIsContextClick = true;

        final ContextBottomDialog.State state = new ContextBottomDialog.State();

        state.mTitle = getItemClickedTitle();
        state.mInfo = TimeUtils.formatDate(getItemClickedDate());
        state.mIsFolder = item instanceof Folder;
        state.mIsWebDav = true;
        state.mIsTrash = isTrash;
        if (!isClickedItemFile()) {
            state.mIconResId = R.drawable.ic_type_folder;
        } else {
            state.mIconResId = getIconContext(StringUtils.getExtensionFromPath(getItemClickedTitle()));
        }
        state.mIsPdf = isPdf();
        if (state.mIsShared && state.mIsFolder) {
            state.mIconResId = R.drawable.ic_type_folder_shared;
        }
        getViewState().onItemContext(state);
    }

    @Override
    public void onActionClick() {
        getViewState().onActionDialog();
    }

    @Override
    public boolean move() {
        if (super.move()) {
            transfer(Api.Operation.DUPLICATE, true);
            return true;
        } else {
            return false;
        }
    }

    @Override
    public boolean copy() {
        if (super.copy()) {
            transfer(Api.Operation.DUPLICATE, false);
            return true;
        } else  {
            return false;
        }
    }

    @Override
    public void upload(@Nullable Uri uri, @Nullable ClipData uris) {
        if (mPreferenceTool.getUploadWifiState() && !NetworkUtils.isWifiEnable(mContext)) {
            getViewState().onSnackBar(mContext.getString(R.string.upload_error_wifi));
            return;
        }

        final String id = mModelExplorerStack.getCurrentId();
        if (id != null) {
            List<Uri> uploadUris = new ArrayList<>();

            if (uri != null) {
                uploadUris.add(uri);
            }

            if (uris != null && uris.getItemCount() > 0) {
                for (int i = 0; i < uris.getItemCount(); i++) {
                    uploadUris.add(uris.getItemAt(i).getUri());
                }
            }

            uploadWebDav(id, uploadUris);
        }
    }

    @Override
    public void uploadToMy(@Nullable Uri uri) {
        if (uri != null) {
            if (mPreferenceTool.getUploadWifiState() && !NetworkUtils.isWifiEnable(mContext)) {
                getViewState().onSnackBar(mContext.getString(R.string.upload_error_wifi));
                return;
            }
            if (ContentResolverUtils.getSize(mContext, uri) > FileUtils.STRICT_SIZE) {
                getViewState().onSnackBar(mContext.getString(R.string.upload_manager_error_file_size));
                return;
            }

            AccountsSqlData sqlData = mAccountSqlTool.getAccountOnline();
            if (sqlData != null && sqlData.isWebDav()) {
                uploadWebDav(sqlData.getWebDavPath(), Collections.singletonList(uri));
            }
        }
    }

    @SuppressLint("MissingPermission")
    private void uploadWebDav(String id, List<Uri> uriList) {
        if (id.charAt(id.length() - 1) != '/') {
            id = id + "/";
        }
        showDialogWaiting(TAG_DIALOG_CANCEL_UPLOAD);
        mUploadDisposable = mFileProvider.upload(id, uriList)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(integer -> { }, throwable -> {
                            fetchError(throwable);
                            if (mTempFile != null && !mTempFile.getWebUrl().equals("")) {
                                FileUtils.asyncDeletePath(Uri.parse(mTempFile.getWebUrl()).getPath());
                            }
                        },
                        () -> {
                            deleteTempFile();
                            getViewState().onDialogClose();
                            getViewState().onSnackBar(mContext.getString(R.string.upload_manager_complete));
                            ((WebDavFileProvider) mFileProvider).getUploadFile().clear();
                        });
    }

    @Override
    public void terminate() {
        super.terminate();
        if (mDownloadDisposable != null) {
            mDownloadDisposable.dispose();
        }
    }

    public void deleteTempFile() {
        if (mTempFile != null && PermissionUtils.checkReadWritePermission(mContext)) {
            Uri uri = Uri.parse(mTempFile.getWebUrl());
            if (uri.getPath() != null) {
                FileUtils.asyncDeletePath(uri.getPath());
            }
        }
    }

    private Explorer removeVideo(Explorer listMedia) {
        List<File> files = new ArrayList<>();
        for (File file : listMedia.getFiles()) {
            if (StringUtils.isImage(file.getFileExst())) {
                files.add(file);
            }
        }
        listMedia.setFiles(files);
        return listMedia;
    }
}

File: app_manager/appmanager/src/main/java/app/editors/manager/mvp/presenters/main/DocsWebDavPresenter.kt
package app.editors.manager.mvp.presenters.main

import android.annotation.SuppressLint
import android.content.Context
import android.net.Uri
import android.os.Build
import androidx.core.net.toUri
import app.documents.core.model.cloud.PortalProvider
import app.documents.core.model.cloud.Recent
import app.documents.core.model.cloud.WebdavProvider
import app.documents.core.network.common.contracts.ApiContract
import app.documents.core.network.manager.models.explorer.CloudFile
import app.documents.core.network.manager.models.explorer.Explorer
import app.documents.core.network.manager.models.request.RequestDownload
import app.documents.core.providers.WebDavFileProvider
import app.editors.manager.R
import app.editors.manager.app.App
import app.editors.manager.app.accountOnline
import app.editors.manager.managers.works.BaseDownloadWork
import app.editors.manager.managers.works.WebDavDownloadWork
import app.editors.manager.mvp.views.main.DocsWebDavView
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import lib.toolkit.base.managers.receivers.ExportReceiver
import lib.toolkit.base.managers.receivers.ExportReceiver.OnExportFile
import lib.toolkit.base.managers.utils.ContentResolverUtils.getSize
import lib.toolkit.base.managers.utils.EditType
import lib.toolkit.base.managers.utils.FileUtils
import lib.toolkit.base.managers.utils.FileUtils.asyncDeletePath
import lib.toolkit.base.managers.utils.NetworkUtils.isWifiEnable
import lib.toolkit.base.managers.utils.PermissionUtils.checkReadWritePermission
import lib.toolkit.base.managers.utils.StringUtils
import moxy.InjectViewState
import retrofit2.HttpException


@InjectViewState
class DocsWebDavPresenter : DocsBasePresenter<DocsWebDavView, WebDavFileProvider>() {

    companion object {
        val TAG: String = DocsWebDavPresenter::class.java.simpleName
    }


    private val exportReceiver: ExportReceiver = ExportReceiver()
    private var tempFile: CloudFile? = null

    init {
        App.getApp().appComponent.inject(this)
    }

    fun getItems() {
        val path = (context.accountOnline?.portal?.provider as? PortalProvider.Webdav)?.path
        getItemsById(path)
    }

    @SuppressLint("UnspecifiedRegisterReceiverFlag")
    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        exportReceiver.onExportReceiver = object : OnExportFile {
            override fun onExportFile(uri: Uri) {
                upload(uri, null)
            }
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            context.registerReceiver(exportReceiver, ExportReceiver.getFilters(), Context.RECEIVER_EXPORTED)
        } else {
            context.registerReceiver(exportReceiver, ExportReceiver.getFilters())
        }

    }

    override fun onDestroy() {
        super.onDestroy()
        exportReceiver.onExportReceiver = null
        try {
            context.unregisterReceiver(exportReceiver)
        } catch (error: IllegalArgumentException) {
            error.printStackTrace()
        }
    }

    override fun getNextList() {
        // Stub
    }

    override fun fetchError(throwable: Throwable) {
        App.getApp().oneDriveComponent
        context.accountOnline?.let { account ->
            val webdavProvider = (account.portal.provider as? PortalProvider.Webdav)?.provider
            val config = App.getApp().appComponent.owncloudTokenDataSource.config
            if (webdavProvider == WebdavProvider.OwnCloud && checkAuthException(throwable) && config != null) {
                viewState.onOwnCloudAuthorization(config)
            } else {
                super.fetchError(throwable)
            }
        } ?: run {
            super.fetchError(throwable)
        }
    }

    override fun cloudFileToRecent(cloudFile: CloudFile): Recent {
        return super.cloudFileToRecent(cloudFile)
            .copy(path = modelExplorerStack.currentId.orEmpty(), isWebdav = true)
    }

    override fun updateViewsState() {
        if (isSelectionMode) {
            viewState.onStateUpdateSelection(true)
            viewState.onActionBarTitle(modelExplorerStack.countSelectedItems.toString())
            viewState.onStateAdapterRoot(modelExplorerStack.isNavigationRoot)
            viewState.onStateActionButton(false)
        } else if (isFilteringMode) {
            viewState.onActionBarTitle(context.getString(R.string.toolbar_menu_search_result))
            viewState.onStateUpdateFilter(true, filteringValue)
            viewState.onStateAdapterRoot(modelExplorerStack.isNavigationRoot)
            viewState.onStateActionButton(false)
        } else if (!modelExplorerStack.isRoot) {
            viewState.onStateAdapterRoot(false)
            viewState.onStateUpdateRoot(false)
            viewState.onStateActionButton(true)
            viewState.onActionBarTitle(currentTitle)
        } else {
            if (pickerMode == PickerMode.Folders) {
                viewState.onActionBarTitle(context.getString(R.string.operation_title))
                viewState.onStateActionButton(false)
            } else {
                viewState.onActionBarTitle("")
                viewState.onStateActionButton(true)
            }
            viewState.onStateAdapterRoot(true)
            viewState.onStateUpdateRoot(true)
        }
    }

    override fun onActionClick() {
        viewState.onActionDialog()
    }

    override fun move(): Boolean {
        return if (super.move()) {
            transfer(ApiContract.Operation.DUPLICATE, true)
            true
        } else {
            false
        }
    }

    override fun copy(): Boolean {
        return if (super.copy()) {
            transfer(ApiContract.Operation.DUPLICATE, false)
            true
        } else {
            false
        }
    }

    override fun upload(uri: Uri?, uris: List<Uri>?, tag: String?) {
        if (preferenceTool.uploadWifiState && !isWifiEnable(context)) {
            viewState.onSnackBar(context.getString(R.string.upload_error_wifi))
            return
        }
        val id = modelExplorerStack.currentId
        if (id != null) {
            val uploadUris: MutableList<Uri> = ArrayList()
            if (uri != null) {
                uploadUris.add(uri)
            }
            if (!uris.isNullOrEmpty()) {
                for (i in uris.indices) {
                    uploadUris.add(uris[i])
                }
            }
            uploadWebDav(id, uploadUris)
        }
    }

    override fun uploadToMy(uri: Uri) {
        if (preferenceTool.uploadWifiState && !isWifiEnable(context)) {
            viewState.onSnackBar(context.getString(R.string.upload_error_wifi))
            return
        }
        if (getSize(context, uri) > FileUtils.STRICT_SIZE) {
            viewState.onSnackBar(context.getString(R.string.upload_manager_error_file_size))
            return
        }
        CoroutineScope(Dispatchers.Default).launch {
            context.accountOnline?.let {
                val webDavProvider = it.portal.provider
                if (webDavProvider is PortalProvider.Webdav) {
                    withContext(Dispatchers.Main) {
                        uploadWebDav(webDavProvider.path, listOf(uri))
                    }
                }
            }
        }
    }

    override fun startDownloadWork(
        to: Uri,
        id: String?,
        url: String?,
        requestDownload: RequestDownload?,
        worker: Class<out BaseDownloadWork>
    ) {
        super.startDownloadWork(to, id, url, requestDownload, WebDavDownloadWork::class.java)
    }

    @SuppressLint("MissingPermission")
    private fun uploadWebDav(id: String, uriList: List<Uri>) {
        var uploadId = id
        if (uploadId[uploadId.length - 1] != '/') {
            uploadId = "$uploadId/"
        }
        uploadDisposable = fileProvider.upload(uploadId, uriList)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({ }, { throwable: Throwable ->
                fetchError(throwable)
                if (tempFile != null && tempFile!!.webUrl != "") {
                    asyncDeletePath(tempFile?.webUrl?.toUri()?.path.orEmpty())
                }
            }
            ) {
                refresh()
                deleteTempFile()
                viewState.onDialogClose()
                viewState.onSnackBar(context.getString(R.string.upload_manager_complete))
                fileProvider.uploadsFile.clear()
            }
    }

    override fun terminate() {
        super.terminate()
        if (downloadDisposable != null) {
            downloadDisposable!!.dispose()
        }
    }

    @SuppressLint("MissingPermission")
    fun deleteTempFile() {
        if (tempFile != null && checkReadWritePermission(context)) {
            val uri = tempFile?.webUrl?.toUri()
            if (uri?.path != null) {
                asyncDeletePath(uri.path.orEmpty())
            }
        }
    }

    private fun removeVideo(listMedia: Explorer): Explorer {
        val files: MutableList<CloudFile> = ArrayList()
        for (file in listMedia.files) {
            if (StringUtils.isImage(file.fileExst)) {
                files.add(file)
            }
        }
        listMedia.files = files
        return listMedia
    }

    override fun createDocs(title: String) {
        modelExplorerStack.currentId?.let { folderId ->
            disposable.add(
                fileProvider.createFile(
                    folderId = folderId,
                    title = title
                )
                    .doOnNext { file ->
                        viewState.onDialogClose()
                        viewState.onOpenLocalFile(file, EditType.Edit(false))
                    }
                    .doOnError {
                        viewState.onDialogClose()
                        viewState.onError(context.getString(R.string.errors_create_local_file))
                    }
                    .subscribe()
            )
        }
    }

    override fun updateDocument(id: String, uri: Uri) {
        downloadDisposable =
            fileProvider.upload(modelExplorerStack.currentId.orEmpty(), listOf(uri))
                .doOnError(::fetchError)
                .doOnNext {
                    refresh()
                    deleteTempFile()
                }
                .subscribe()
    }

    private fun checkAuthException(e: Throwable) : Boolean {
        return e is HttpException && e.code() == ApiContract.HttpCodes.CLIENT_UNAUTHORIZED
    }
}